pkg/spanconfig/spanconfigsqlwatcher/buffer.go

```diff
@@ -0,0 +1,121 @@
+// maintains the notion of the combined frontier timestamp computed as the
+// minimum of the two. This is used when flushing the buffer periodically.
+type buffer struct {
+	mu struct {
+		syncutil.Mutex
+
+		// rangefeed.Buffer stores spanconfigsqlwatcher.Events.
+		buffer *rangefeedbuffer.Buffer
+
+		// rangefeedFrontiers tracks the frontier timestamps of individual
+		// rangefeeds established by the SQLWatcher.
+		rangefeedFrontiers [numRangefeeds]hlc.Timestamp
+	}
+}
+
+// event is the unit produced by the rangefeeds the SQLWatcher establishes over
+// system.zones and system.descriptors. It implements the rangefeedbuffer.Event
+// interface.
+type event struct {
+	// timestamp at which the event was generated by the rangefeed.
+	timestamp hlc.Timestamp
+
+	// update captures information about the descriptor or zone that the
+	// SQLWatcher has observed change.
+	update spanconfig.SQLWatcherUpdate
+}
+
+// Timestamp implements the rangefeedbuffer.Event interface.
+func (e event) Timestamp() hlc.Timestamp {
+	return e.timestamp
+}
+
+// rangefeedKind is used to identify the distinct rangefeeds {descriptors,
+// zones} established by the SQLWatcher.
+type rangefeedKind int
+
+const (
+	zonesRangefeed rangefeedKind = iota
+	descriptorsRangefeed
+
+	// numRangefeeds should be listed last.
+	numRangefeeds
+)
+
+// newBuffer constructs and returns a new buffer.
+func newBuffer(limit int) *buffer {
+	rangefeedBuffer := rangefeedbuffer.New(limit)
+	eventBuffer := &buffer{}
+	eventBuffer.mu.buffer = rangefeedBuffer
+	return eventBuffer
+}
+
+// advance advances the frontier for the given rangefeed.
+func (b *buffer) advance(rangefeed rangefeedKind, timestamp hlc.Timestamp) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.mu.rangefeedFrontiers[rangefeed].Forward(timestamp)
+}
+
+// add records the given event in the buffer.
+func (b *buffer) add(ev event) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	return b.mu.buffer.Add(ev)
+}
+
+// flush computes the combined frontier timestamp of the buffer and returns a
+// list of unique spanconfig.SQLWatcherUpdates below this timestamp. The
+// combined frontier timestamp is also returned.
+func (b *buffer) flush(
+	ctx context.Context,
+) (updates []spanconfig.SQLWatcherUpdate, combinedFrontierTS hlc.Timestamp) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	seenIDs := make(map[descpb.ID]struct{})
+	// First we determine the checkpoint timestamp, which is the minimum
+	// checkpoint timestamp of all event types.
+	combinedFrontierTS = hlc.MaxTimestamp
+	for _, ts := range b.mu.rangefeedFrontiers {
+		combinedFrontierTS.Backward(ts)
+	}
+
+	events := b.mu.buffer.Flush(ctx, combinedFrontierTS)
+
+	for _, ev := range events {
+		update := ev.(event).update
+		// De-duplicate IDs from the returned result.
+		if _, seen := seenIDs[update.ID]; !seen {
+			seenIDs[update.ID] = struct{}{}
+			updates = append(updates, update)
+		}
+	}
+	return updates, combinedFrontierTS
```

Comment:
I can't tell you why I care about this, because it's not really reasonable. It's totally fine to allocate the memory you're allocating here and what not. Nevertheless, I feel like typing these nitpicks, so here you go:

1) Let's only hold the mutex on the buffer while we find the timestamp:

```go
// flush computes the combined frontier timestamp of the buffer and returns a
// the relevant events which were buffered up to that timestamp.
func  flushEvents(
	ctx context.Context,
) (updates []rangefeedbuffer.Event, combinedFrontierTS hlc.Timestamp) {
         b.mu.Lock()
	defer b.mu.Unlock()
	// First we determine the checkpoint timestamp, which is the minimum
	// checkpoint timestamp of all event types.
	combinedFrontierTS = hlc.MaxTimestamp
	for _, ts := range b.mu.rangefeedFrontiers {
		combinedFrontierTS.Backward(ts)
	}

	return b.mu.buffer.Flush(ctx, combinedFrontierTS), combinedFrontierTS
}
```
2) more nit-picky, we don't need the map
```go
// flush computes the combined frontier timestamp of the buffer and returns a
// list of unique spanconfig.SQLWatcherUpdates below this timestamp. The
// combined frontier timestamp is also returned.
func (b *buffer) flush(
	ctx context.Context,
) (updates []spanconfig.SQLWatcherUpdate, combinedFrontierTS hlc.Timestamp) {
        var events []rangefeedbuffer.Event
        events, combinedFrontierTS = b.flushEvents(ctx)
        sort.Slice(events, func(i, j int) bool {
           ei, ej := events[i].(event), events[j].(event)
           if ei.update.ID == ej.update.ID {
               return ei.timestamp.Less(ej.timestamp)
           }
           return ei.update.ID < ej.update.ID
        })
        for i, ev := range events {
           if i == 0 || events[i-1].(event).update.ID != ev.(event).update.ID {
              updates = append(updates, ev.(event).update)
           }
        }
	return updates, combinedFrontierTS
}
```